### ISSUE TITLE
Extending Arrays misses error detail

### ISSUE THREAD

**Issue by joseandrespg** [sylvainpolletvillard/ObjectModel#37]
Hi,

We are trying to use the extend operation to encapsulate some validation logic into helper classes, but we are also finding some issues with the errors reported.

We tried to different approaches:

- Case 1:
We miss the root error and we get the parent element instead of the invalid property.

- Case 2:
We get the root error, but we miss the path and the extra info about what was wrong.

Here an example:

```javascript
var debug = require('debug')('log');
var Model = require("objectmodel");

function printErrors(errors) {
    console.log("Caught these errors:");
    errors.forEach(function(error){ debug(error); });
}

function assertMandatoryCollection(collection) {
    return collection && Array.isArray(collection) && collection.length > 0
}

var MandatoryCollection = new Model.Array()
.assert(assertMandatoryCollection, 'MandatoryCollection should have at least one child');

var CountryCode = new Model(String)
.assert(function (a) {
    return a === "GB";
}, "Country must be GB");

var Carrier = Model.Object({
    code: [String],
    service: [String],
    alias: [String]
})
.assert(function (c) {
    return !(!c.alias && (!c.code || !c.service));
}, 'Should indicate code and service or indicate an alias');

var Address = new Model({
    city: String,
    country: CountryCode,
    carrier: Carrier
});

// Case 1, using extend()
var addressesExtended = Model.Array(Address)
.extend(MandatoryCollection);

var Order1 = new Model({
    sku: String,
    addresses: addressesExtended
});

// Case 2, using assert()
var addressesAsserted = Model.Array(Address)
.assert(assertMandatoryCollection, 'Addresses should have at least one child');

var Order2 = new Model({
    sku: String,
    addresses: addressesAsserted
});

// Examples
var gbAddress = { city: "London", country: "GB", carrier: {alias: "alias"} }; // valid address
var frAddress = { city: "Paris", country: "FR", carrier: {code: "code"} }; // invalid carrier, missing service property

var gbOrder = { sku: "ABC123", addresses: [gbAddress] };
var frOrder = { sku: "ABC123", addresses: [frAddress] };

Order1.validate(gbOrder, printErrors); // no errors
Order1.validate(frOrder, printErrors); // error shows parent element

Order2.validate(gbOrder, printErrors); // no errors
Order2.validate(frOrder, printErrors); // missing errors path
```

And this is the console output:

```
Caught these errors:
  log { expected: 
  log    [ { [Function]
  log        definition: [Object],
  log        assertions: [],
  log        errorCollector: [Function] },
  log      { [Function]
  log        definition: undefined,
  log        assertions: [Object],
  log        errorCollector: [Function] } ],
  log   received: { city: 'Paris', country: 'FR', carrier: { code: 'code' } },
  log   path: 'addresses[0]',
  log   message: 'expecting addresses[0] to be {\n\tcity: String,\n\tcountry: String,\n\tcarrier: {\n\t\tcode: [String],\n\t\tservice: [String],\n\t\talias: [String]\n\t}\n} or Array of undefined, got Object {\n\tcity: "Paris",\n\tcountry: "FR",\n\tcarrier: {\n\t\tcode: "code"\n\t}\n}' } +0ms
Caught these errors:
  log { message: 'assertion "Country must be GB" returned false for value "FR"' } +6ms
  log { message: 'assertion "Should indicate code and service or indicate an alias" returned false for value {\n\tcode: "code"\n}' } +1ms
```

Which is the right way to do this inheritance?

Are we missing something?

Thanks!

**Comment by sylvainpolletvillard** [comment]
Hello !

I'm not sure I fully understand your example and your expectations. If you have several different questions, it would be easier to split them in several issues instead of an all-in-one example.

In your example, MandatoryCollection is defined as an Array of undefined values. Indeed, `Model.Array()` takes the definition as argument and since no argument has been passed, the definition is `undefined`.

```
MandatoryCollection.toString()
> "Array of undefined"
```

I think you expected MandatoryCollection to validate **any** type as array elements, but currently there is no such **any** type in ObjectModel. Through you can add it easily:

```
Model.Any = Model([Object, Boolean, Number, String, Symbol, undefined, null]);
var MandatoryCollection = Model.Array(Model.Any)
```

If you extend MandatoryCollection, the definition will be extended too. So `MandatoryCollection.extend(Address)` is probably not what you want, because it will still accept **any** type as array items and not only Address.

For this kind of "assertion-only model", I think the `extend `method is not appropriate since you don't want to extend the MandatoryCollection definition, only the assertions. So I would do it this way:
```
var Addresses = Model.Array(Address);
Addresses.assertions = Addresses.assertions.concat(MandatoryCollection.assertions);
```

Concerning the details of errors in nested sub-model validation, this is a known issue that I am working on for next version of ObjectModle. This is not easy to provide the right level of information that always helps the developper fixing its code. Going too deep in the error level sometimes makes things even more confusing. 

Could you please open another issue for the lack of information in nested sub-model errors, with a single example and your expectation ? This issue would be restricted to usage of extend as it is your first question.

**Comment by joseandrespg** [comment]
Thanks for your quick response!

I see. `Model.Array` is not working as we were expecting. Understood.

I'm closing this issue and will open a new one for the missing error extra info.

Cheers.

**Comment by sylvainpolletvillard** [comment]
I'm going to make a few changes in the future to avoid this confusion:
* make the definition argument mandatory in model constructors
* make `assert` method accept an array to simplify to the code below into
 `var Addresses = Model.Array(Address).assert(MandatoryCollection.assertions);`

Let me know how docs can be improved too.

**Comment by joseandrespg** [comment]
That changes will definitely help.

I would add a point inside the 'Array models' section, once you get the above changes:

**Extending array models validations**

``` javascript
var Cards = Model.Array([Number, "J","Q","K"]);
var ArrayUnique = Model.Array(Model.Any).assert(function hasNoDuplicates(arr){
    return arr.every(function(val, idx){
        return arr.indexOf(val) === idx
    })
});

var UniqueCards = Cards.extend(ArrayUnique); // won't work, UniqueCards will allow any types

var UniqueCards = Cards.extend().assert(ArrayUnique.assertions); // use assert instead

var cardCollection = new UniqueCards([1,2,3,4]);

cardCollection.push(5); // no errors
cardCollection.push(1); // TypeError: assertion "hasNoDuplicates" returned false for value [1, 2, 3, 4, 5, 1]
```

**Comment by sylvainpolletvillard** [comment]
I changed my mind about the `assert` API change since it does not play well with the second argument `description`. Also there are plenty of short ways to do it without requiring an API change
* `Addresses.assertions = Addresses.assertions.concat(MandatoryCollection.assertions)`
* `Addresses.assertions = MandatoryCollection.assertions.slice()`
* `Addresses.assertions.push(...MandatoryCollection.assertions)`

Basically any way to clone the parent model assertions array would do the trick. Of course you can also refer to the `hasNoDuplicates` function instead of extending a model, especially if this model has no useful definition.

I will update the documentation on `extend` since it does not explain the merge strategy apart from object models.
